[ICD]Trigger resubscription when receiving check-in and subscription has not yet become abnormal

yunhanw-google opened the pull request:
If the server sends out check-in message, and there is no reschedule subscription yet in client side at the same time, it means current client does not realize subscription has gone, and we should forcibly timeout current subscription, and schedule a new one.

fixes https://github.com/project-chip/connectedhomeip/issues/30902


#### Testing
Drafting

yunhanw-google committed:
Trigger resubscription when receiving check-in and subscription has not yet become abnoaml in client

yunhanw-google committed:
fix OnPeerTypeChange

--When sit transition to lit, check-in would be sent and activate the subscription. if any client subscribe to operating mode, and detect it becoems sit,
just update the operating mode, there is no need to re-activate the subscription.

bzbarsky-apple commented on src/app/ReadClient.cpp:
```diff
@@ -484,6 +484,7 @@ CHIP_ERROR ReadClient::GenerateDataVersionFilterList(DataVersionFilterIBs::Build
 
 void ReadClient::OnActiveModeNotification()
 {
+    // Note: this API only works when issuing subscription via SendAutoResubscribeRequest.
```
So hold on.  When not doing auto-resubscribe, do we want to ignore this call, or still close ourselves even though it's fatal?

yunhanw-google replied:
I think the whole OnActiveModeNotification is designed for auto-resubscription stuff, if they are not using OnActiveModeNotification, it is really application consumer's responsibility to decide how to manipulate close and retry logic when receiving check-in message, and application consumer need implement their CheckInDelegate and not call InteracitonModelEngine::OnActiveModeNotification.

I think we should ignore this call when not using auto-resubscription.

yunhanw-google replied:
just update code to reflect this change.

bzbarsky-apple commented on src/app/ReadClient.cpp:
```diff
@@ -493,6 +494,18 @@ void ReadClient::OnActiveModeNotification()
         return;
     }
 
+    // If the server sends out check-in message, and there is no reschedule subscription yet in client side at the same time, it means
+    // current client does not realize subscription has gone, and we should forcibly timeout current subscription, and schedule a new one.
```
So the thing is: per spec, the check-in target and the thing whose subscription the server checks for might be different nodes.  And in practice this is actually going to happen, I can guarantee that.

So it's entirely possible for this subscription to be _just fine_ and yet for a check-in message to arrive because some _other_ subscription from a different node is not fine.

I assume we're already surfacing check-in messages to the SDK API consumer?  If so, they would need to retrigger subscriptions as needed.  Or we need some way to configure this behavior, at the very least, because the right thing to do here depends on information this code does not have.

yunhanw-google replied:
yes, we have already surfacing check-in messages to the SDK API consumer, it is consumer's reponsibility to retrigger subscription as need, they can mimic OnActiveModeNotification behavior to configure their behavior.

yunhanw-google replied:
> So the thing is: per spec, the check-in target and the thing whose subscription the server checks for might be different nodes. And in practice this is actually going to happen, I can guarantee that.

If subscription from node A is and good, register node for check-in is for nodeB,  check-in message would never arrive at nodeA, right? since server send check-in message based upon registered node, unless we allow that multiple nodes reside in same client device? maybe we need add one more check between ICDClientInfo::check_in_node against the node id extracted from readClient in InteractionModelEngine:: OnActiveModeNotification?

yunhanw-google replied:
just updated code to reflect the above change.

bzbarsky-apple replied:
> If subscription from node A is and good, register node for check-in is for nodeB, check-in message would never arrive at nodeA, right?

Wrong scenario.  The scenario is that node A is check-in target and has a subscription but that is NOT the subscription check-in is tracking.  Node B is the one the check-in is tracking and that one is dead.  Node A will receive the check-in, because Node B is not subscribed, and the right action is to prod Node B (or maybe Node C, depending) in the app, not tear down A's perfectly OK subscription.

Point being: ReadClient does not have the information to make this decision; the application does.

yunhanw-google committed:
address comments

yunhanw-google committed:
add monitor subject check for subscription activation

If caseTag for current subscription does not match with the one set in ICDManagementCluster::RegisterClient for check-in,
when receiving check-in message, OnActiveModeNotification would do nothing for current subscription.

yunhanw-google committed:
fix compilation

bzbarsky-apple reviewed (changes requested):
Requesting changes because the code does not make sense and clearly needs testing.

yunhanw-google committed:
introduce mCatsMatchCheckIn to indicate whether cats for icd registration matched with the current subscription

yunhanw-google commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -1069,14 +1069,18 @@ void InteractionModelEngine::OnResponseTimeout(Messaging::ExchangeContext * ec)
+        // If caseTag for current subscription does not match with the one set in ICDManagementCluster::RegisterClient for check-in,
+        // when receiving check-in message, OnActiveModeNotification would do nothing for current subscription.
+        if (ScopedNodeId(pListItem->GetPeerNodeId(), pListItem->GetFabricIndex()) == aPeer &&
+           pListItem->GetSubjectDescriptor().HasValue() && pListItem->GetSubjectDescriptor().cats.CheckSubjectAgainstCATs(aMonitoredSubject) &&
```
yep, this does not work, i removed this change, and introduce one parameter via ReadPrepareParam from application, indicate current subscription's cats  is not same as cats for check-in's registration...

bzbarsky-apple commented on src/app/ReadClient.cpp:
```diff
@@ -485,6 +485,15 @@ CHIP_ERROR ReadClient::GenerateDataVersionFilterList(DataVersionFilterIBs::Build
 {
     VerifyOrDie(mpImEngine->InActiveReadClientList(this));
+
+    // Note: this API only works when issuing subscription via SendAutoResubscribeRequest, when SendAutoResubscribeRequest is
```
```suggestion
    // Note: this API only works when issuing subscription via SendAutoResubscribeRequest. When SendAutoResubscribeRequest is
```

yunhanw-google replied:
updated, thanks

bzbarsky-apple commented on src/app/ReadClient.cpp:
```diff
@@ -485,6 +485,15 @@ CHIP_ERROR ReadClient::GenerateDataVersionFilterList(DataVersionFilterIBs::Build
+
+    // Note: this API only works when issuing subscription via SendAutoResubscribeRequest, when SendAutoResubscribeRequest is
+    // called, either mEventPathParamsListSize or mAttributePathParamsListSize is not 0.
+    VerifyOrDie(mReadPrepareParams.mEventPathParamsListSize != 0 || mReadPrepareParams.mAttributePathParamsListSize != 0);
```
Hold on a sec.  Why is this a VerifyOrDie?  Why should we _crash_ if we happen to have some random subscription that is not supposed to be auto-resubscribed?

Stopping here; please make the logic in this PR actually make sense before I spend more time looking at it.

yunhanw-google replied:
yes, i agree, if we have random subscription without auto, we should do nothing here.

yunhanw-google replied:
update with VerifyOrReturn here, thanks

yunhanw-google committed:
address comments and add tests

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

bzbarsky-apple commented on src/app/ReadClient.cpp:
```diff
@@ -485,6 +485,15 @@ CHIP_ERROR ReadClient::GenerateDataVersionFilterList(DataVersionFilterIBs::Build
+    // called, either mEventPathParamsListSize or mAttributePathParamsListSize is not 0.
+    VerifyOrReturn(mReadPrepareParams.mEventPathParamsListSize != 0 || mReadPrepareParams.mAttributePathParamsListSize != 0);
+
+    // If mCatsMatchCheckIn is true, it means cats used in icd registration matches with the one in current subscription,
```
Matches _which side_?  Should be "our side of the current subscription" or something, right?

But do we really want to push this off onto the API client?  Or just examine the fabric table?

And if we do push this off onto the API client, as this PR does, then should this boolean just be `mResetSubscriptionOnCheckin`?

yunhanw-google replied:
If we are going with fabrictable path, then we can get fabricTable from interaction model engine, then use get fabricId via FabricIndex readClient::GetFabricIndex() const { return mPeer.GetFabricIndex(); },  further we can get cats via FetchCATs(mFabricIndex, cats).

yunhanw-google replied:
this logic has been removed, we are leaning toward examining the fabric table, thanks

bzbarsky-apple commented on src/app/ReadClient.cpp:
```diff
@@ -503,12 +525,6 @@ void ReadClient::OnPeerTypeChange(PeerType aType)
-    // If the peer is no longer LIT, try to wake up the subscription and do resubscribe when necessary.
-    if (!mIsPeerLIT)
-    {
-        OnActiveModeNotification();
```
So I am trying to understand why it's OK to remove this.  The reason this seems to be here is that it's envisioning us having two subscriptions, one of which (call it A) is watching for peer type change and one of which (call it B) is IsInactiveICDSubscription().  When the transition to SIT is observed, B will no longer get any OnActiveModeNotifications, so how will it get out of the "inactive ICD subscription" state?

Of course if A does not exist and we just have B, and it's in the "inactive ICD subscription" state and then the server changes to A SIT, I think we just get stuck; that seems to be a hole in the whole logic of this stuff, including the spec?

yunhanw-google replied:
The scenario is as below
Assume we have one inactive wildcard subscription between controller and ICD device. When ICD transitions from LIT to SIT, check-in message is sent to controller, and controller activate the sleepy subscription, right? then the operation mode attribute is changed, OnPeerTypeChange is called so that OnActiveModeNotification is called again and this subscription would be recreated.

yunhanw-google replied:
You are right, the current change does not work on the below scenario

> So I am trying to understand why it's OK to remove this. The reason this seems to be here is that it's envisioning us having two subscriptions, one of which (call it A) is watching for peer type change and one of which (call it B) is IsInactiveICDSubscription(). When the transition to SIT is observed, B will no longer get any OnActiveModeNotifications, so how will it get out of the "inactive ICD subscription" state?

yunhanw-google replied:
updated the code.
1. Update PeerTypeChange and call TriggerResubscriptionForLivenessTimeout when peer type is changed to LIT and state is inactive icd subscription, instead of  OnActiveModeNotification to help situation 1 in your question.
2. If icd transition from lit to sit, it would send check-in message, which would bring up the in active subscription.

bzbarsky-apple commented on src/app/ReadClient.h:
```diff
@@ -351,11 +351,17 @@ class ReadClient : public Messaging::ExchangeDelegate
-     * call this function when a check-in message is received.
+     *  call this function when a check-in message is received.
+     *
+     *  If the server sends out check-in message, and there is a active tracked active subscription in client side at the same time,
+     * it means current client does not realize this tracked subscription has gone, and we should forcibly timeout current
+     * subscription, and schedule a new one.
```
This comment does not match what the API actually does, really.

yunhanw-google replied:
rewrite the comments, thanks

yunhanw-google committed:
address comments

bzbarsky-apple commented on src/app/InteractionModelEngine.h:
```diff
@@ -236,16 +236,15 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
      *
      *  ReadClient will call this function when it finds any updates of the OperatingMode attribute from ICD management
-     * cluster. The application doesn't need to call this function, usually.
+     *  cluster. The application doesn't need to call this function, usually.
```
```suggestion
     * cluster. The application doesn't need to call this function, usually.
```

bzbarsky-apple commented on src/app/ReadClient.h:
```diff
@@ -351,11 +351,17 @@ class ReadClient : public Messaging::ExchangeDelegate
-     * call this function when a check-in message is received.
+     *  call this function when a check-in message is received.
+     *
+     *  If the server sends out check-in message, and there is a active tracked active subscription in client side at the same time,
```
There's a lot of verbiage here, but it's all very confusing and mostly describes implementation details not API.

The API is: This function should be called when the peer is an ICD that is checking in and this ReadClient represents a subscription that would cause that ICD to not need to check in anymore.

no?  That's about all the documentation this function needs.

yunhanw-google replied:
updated, thanks

bzbarsky-apple commented on src/app/ReadClient.cpp:
```diff
@@ -493,6 +498,18 @@ void ReadClient::OnActiveModeNotification()
         return;
     }
 
+    // When receiving check-in message, it means all tracked subscriptions for this node has gone in server side, if there is a related active subscription
+    // and subscription has not yet rescheduled in client side, we should forcibly timeout the current subscription, and schedule a new one.
```
```suggestion
    // If this API has been called, that means the subscription for this ReadClient is gone
    // on the server side (because otherwise the server would not have checked in with us).
    // Even if we think we have a live subscription, we are wrong, and should just forcibly time it
    // out and schedule a new one.
```

bzbarsky-apple commented on src/app/ReadClient.cpp:
```diff
@@ -493,6 +498,18 @@ void ReadClient::OnActiveModeNotification()
+        return;
+    }
+
+    // If the server sends a check-in message and a subscription is already scheduled, it indicates a client-side subscription
+    // timeout or failure. Cancel the scheduled subscription and initiate a new one immediately.
```
```suggestion
    // If we have already detected subscription loss and are waiting to try to re-subscribe,
    // now is a really good time to do it, since the server is listening.
```

bzbarsky-apple commented on src/app/ReadClient.cpp:
```diff
@@ -504,10 +521,10 @@ void ReadClient::OnPeerTypeChange(PeerType aType)
 
-    // If the peer is no longer LIT, try to wake up the subscription and do resubscribe when necessary.
-    if (!mIsPeerLIT)
+    // If the peer is no longer LIT and in inactive ICD subscription status, try to wake up the subscription and do resubscribe.
```
```suggestion
    // If the peer is no longer LIT and we were waiting for a check-in to try to resubscribe,
    // just try to resubscribe now, because a SIT is not going to send a check-in.
```

yunhanw-google committed:
Update src/app/InteractionModelEngine.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/app/ReadClient.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/app/ReadClient.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/app/ReadClient.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -1069,14 +1069,16 @@ void InteractionModelEngine::OnResponseTimeout(Messaging::ExchangeContext * ec)
-        if (ScopedNodeId(pListItem->GetPeerNodeId(), pListItem->GetFabricIndex()) == aPeer)
+        CATValues cats;
+        mpFabricTable->FetchCATs(pListItem->GetFabricIndex(), cats);
+        if (ScopedNodeId(pListItem->GetPeerNodeId(), pListItem->GetFabricIndex()) == aPeer && cats.CheckSubjectAgainstCATs(aMonitoredSubject))
```
in that situation, I think we need compare the aMonitoredSubject against readClient::getLocalNodeId, if it is equal, we would proceed with OnActiveModeNotification

yunhanw-google committed:
address comments

yunhanw-google commented on src/app/InteractionModelEngine.h:
```diff
@@ -236,16 +236,15 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
      *
-     *  When subscribing to ICD and liveness timeout reached, the read client will move to `InactiveICDSubscription` state and
-     * resubscription can be triggered via OnActiveModeNotification().
+     *  See ReadClient::OnActiveModeNotification
```
updated, thanks

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

yunhanw-google committed:
fix

bzbarsky-apple commented on src/app/ReadClient.h:
```diff
@@ -645,6 +639,7 @@ class ReadClient : public Messaging::ExchangeDelegate
     uint16_t mMaxInterval              = 0;
     SubscriptionId mSubscriptionId     = 0;
     ScopedNodeId mPeer;
+    NodeId mLocal;
```
Why are we bloating these things?  The caller can just get the local node ID from the fabric table, no?

yunhanw-google replied:
agree, removed, just obtain local node id from fabric table. thanks

bzbarsky-apple commented on src/app/InteractionModelEngine.h:
```diff
@@ -234,17 +234,18 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
-     *  Activate the idle subscriptions.
      *
-     *  See ReadClient::OnActiveModeNotification
+     *  When a peer becomes active, it sends a check-in message because aMonitoredSubject does not have a subscription to it.
+     * Upon receiving this check-in message, the Client attempts to run the re-subscription process via OnActiveModeNotification
```
```suggestion
     *  Notification that aPeer has sent a check-in message because aMonitoredSubject does
     *  not have a subscription to it.
```

yunhanw-google committed:
Update src/app/InteractionModelEngine.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
address comments

restyled-commits committed:
Restyled by clang-format

yunhanw-google committed:
fix

yunhanw-google committed:
update tests
